### ISSUE TITLE
About the implementation details of angular margin.

### ISSUE THREAD

**Issue by rocketbear** [deepinsight/insightface#228]
I have read the implementation of angular margin in your code. It seems there are some details not mentioned in the paper.

Let t be the angle between a feature vector and the weight vector of its groundtruth class. The value of t is in [0, pi]. Let m be the margin parameter. m must be < pi/2 and >= 0. Let s be the scaling factor.

If no margin is applied, the score of the groundtruth class input to the softmax layer is s * cos(t).
The angular margin is then implemented as follows:
if t + m < pi, i.e. cos(t) > cos(pi - m), the score of the groundtruth class is changed to s * cos(t+m).
if t + m >= pi: If easy_margin is true, the score of the groundtruth class is not modified. Otherwise the score of the  groundtruth class is changed to s * (cos(t) - m*sin(pi-m)).

So for very hard instances (t + m >= pi), the code actually falls back to cosine margin with margin=m*sin(pi-m). This actually creates a discontinuity point on the target logit curve. 

The paper says:
    the target logit curve is not monotonic decreasing  when theta in [0, 180] ... the increasing interval of ArcFace is almost never reached during training. Therefore we do not need to deal with this explicitly.

But actually you have dealt with this explicitly in the code.

**Comment by Lxyzgf** [comment]
@rocketbear 
I am also confused about this point. Could anyone else  explain this question？

**Comment by nttstar** [comment]
I just want to keep the value strictly decreasing for theta in [0,180]
